src/ripple/app/tx/impl/applySteps.cpp

```diff
@@ -460,173 +315,17 @@ TxConsequences::TxConsequences(STTx const& tx, std::uint32_t sequencesConsumed)
 static std::pair<TER, bool>
 invoke_apply(ApplyContext& ctx)
 {
-    switch (ctx.tx.getTxnType())
+    try
     {
-        case ttACCOUNT_DELETE: {
-            DeleteAccount p(ctx);
-            return p();
-        }
-        case ttACCOUNT_SET: {
-            SetAccount p(ctx);
-            return p();
-        }
-        case ttCHECK_CANCEL: {
-            CancelCheck p(ctx);
-            return p();
-        }
-        case ttCHECK_CASH: {
-            CashCheck p(ctx);
-            return p();
-        }
-        case ttCHECK_CREATE: {
-            CreateCheck p(ctx);
-            return p();
-        }
-        case ttDEPOSIT_PREAUTH: {
-            DepositPreauth p(ctx);
-            return p();
-        }
-        case ttOFFER_CANCEL: {
-            CancelOffer p(ctx);
-            return p();
-        }
-        case ttOFFER_CREATE: {
-            CreateOffer p(ctx);
-            return p();
-        }
-        case ttESCROW_CREATE: {
-            EscrowCreate p(ctx);
-            return p();
-        }
-        case ttESCROW_FINISH: {
-            EscrowFinish p(ctx);
-            return p();
-        }
-        case ttESCROW_CANCEL: {
-            EscrowCancel p(ctx);
-            return p();
-        }
-        case ttPAYCHAN_CLAIM: {
-            PayChanClaim p(ctx);
-            return p();
-        }
-        case ttPAYCHAN_CREATE: {
-            PayChanCreate p(ctx);
-            return p();
-        }
-        case ttPAYCHAN_FUND: {
-            PayChanFund p(ctx);
-            return p();
-        }
-        case ttPAYMENT: {
-            Payment p(ctx);
-            return p();
-        }
-        case ttREGULAR_KEY_SET: {
-            SetRegularKey p(ctx);
-            return p();
-        }
-        case ttSIGNER_LIST_SET: {
-            SetSignerList p(ctx);
+        return with_txn_type(ctx.tx.getTxnType(), [&]<typename T>() {
```

Comment:
Does the developer need to specify `invoke_apply.template operator<Some_Txn_Type>()` the type explicitly like this? 

So you have effectively collected the all the common "functionality" from `invoke_apply`, `invoke_preflight_helper`, `invoke_preclaim` and `invoke_calculateBaseFee` and placed them under `with_txn_type`. Am I understanding this correctly?

Reply:
I may be misunderstanding your question, but `invoke_apply` is called like this: `invoke_apply(ctx);`. It's a regular function and doesn't have an `operator()`. 

`with_txn_type` factors out the `switch` statement that used to live in those four places, and handles the variability between those functions by using a callback. So now when a new transaction is added we need to change one place rather than four.

Reply:
`invoke_apply` is a normal function, it's just doing complex things under the hood.